Add Network Topology plugin docs (#4981)
elastic/observability-docs: docs/en/observability/network-topology-fields.asciidoc

[[network-topology-fields]]
= Network Topology field reference

This page describes every {es} field the <<network-topology,Network Topology plugin>> reads or writes, its ECS compliance status, mapping type, and expected values.

[discrete]
[[network-topology-fields-ecs-status]]
== ECS status key

[cols="1,3"]
|===
| Badge | Meaning

| *Core ECS*
| Defined in the {ecs-ref}/ecs-reference.html[Elastic Common Schema (ECS)] and used as specified.

> | *ECS ext.*

Reviewer: I think this should say "extension"? instead of the contraction

Reviewer: Or maybe this is standard? Feel free to ignore, I just know we usually avoid abbreviations

| Uses an ECS-defined namespace but with values that extend beyond the official spec.

| *Custom*
| No ECS equivalent — SNMP-specific data for which ECS does not define field sets.
|===

[discrete]
[[network-topology-fields-base]]
== Base fields

[cols="2,1,1,3,2"]
|===
| Field | Type | ECS status | Description | Example

| `@timestamp`
| date
| Core ECS
| Document timestamp — when the SNMP poll was collected.
| `2024-01-01T12:00:00.000Z`
|===

[discrete]
[[network-topology-fields-host]]
== `host.*` — Device identity

ECS `host` fields identify the monitored network device.

[cols="2,1,1,3,2"]
|===
| Field | Type | ECS status | Description | Example

| `host.name`
| keyword
| Core ECS
| Device hostname (SNMP `sysName`).
| `hq-core-rtr-01`

| `host.ip`
| ip
| Core ECS
| Primary management IP address.
| `10.1.1.2`

| `host.mac`
| keyword
| Core ECS
| Primary MAC address.
| `aa:bb:cc:dd:ee:01`

| `host.type`
| keyword
| ECS ext.

> | Device category — set explicitly in the {ls} collector config. The ingest pipeline can infer it from `observer.sys_descr` as a fallback when unset.

Reviewer: ```suggestion
| Device category. Set explicitly in the {ls} collector config. The ingest pipeline infers it from `observer.sys_descr` as a fallback when unset.
```
Unless we're going to add details about how the inference works, I think this is simpler.

Reviewer: Otherwise it leaves me wondering, okay it can, but will it?

| `router`, `switch`, `firewall`, `server`, `ap`, `unknown`
|===

[discrete]
[[network-topology-fields-observer]]
== `observer.*` — Device classification

ECS `observer` fields describe the network device as the observed system.

[cols="2,1,1,3,2"]
|===
| Field | Type | ECS status | Description | Example

| `observer.vendor`
| keyword
| Core ECS

> | Vendor name — auto-detected from `observer.sys_descr` by the ingest pipeline.

Reviewer: Lot of em dashes in these definitions. Fine to leave them, but some people hate them now and may start ranting about AI to their colleagues

Author: haha true. I honestly just did a one to one move from the GH repo for this file and didn't modify it at all.

| `Cisco`, `Juniper`, `Palo Alto`, `Arista`, `Fortinet`, `HPE`, `Aruba`

| `observer.sys_descr`
| text
| Core ECS
| Raw SNMP `sysDescr` string — the ingest pipeline uses this to populate `observer.vendor` and `host.type`.
| `Cisco IOS XR Software, ASR 9000 Series Router`

| `observer.os.full`
| keyword
| Core ECS
| Full OS version string.
| `IOS-XR 7.3.2`, `Junos 22.1R1`
|===

[discrete]
[[network-topology-fields-network]]

> == `network.*` — Location and role metadata

Reviewer: This is important, including to the task that the tutorial works on. Maybe worth mentioning these location/role metadata fields in the tutorial so people know how to get their data associated with the right site?

> These fields provide hierarchical location context for the device. The ECS `network` field set covers protocol and transport data; the site, building, and role fields below are justified custom extensions within that namespace.

Reviewer: ```suggestion
These fields provide hierarchical location context for the device. The ECS `network` field set covers protocol and transport data; the site, building, and role fields below extend that namespace.
```

[cols="2,1,1,3,2"]
|===
| Field | Type | ECS status | Description | Example

| `network.site`
| keyword
| ECS ext.
| Site or datacenter identifier. Defaults to `Ungrouped` if absent (set by the ingest pipeline).
| `HQ-DC1`, `Branch-NYC`, `Branch-CHI`

| `network.building`
| keyword
| ECS ext.
| Building within the site.
| `Main`, `Annex`, `Tower-B`

| `network.role`
| keyword
| ECS ext.
| Network tier — used for topology hierarchy. Controls the vertical position of nodes in the topology map.
| `core`, `distribution`, `access`, `server`
|===

[discrete]
[[network-topology-fields-interface]]
== `interface.*` — SNMP interface metrics (Custom)

> No ECS field set covers SNMP interface MIB (IF-MIB, RFC 2863) data. These fields are plugin-defined under a consistent `interface.*` namespace.

Reviewer: ```suggestion
No ECS field set covers SNMP interface MIB (IF-MIB, RFC 2863) data. These fields are plugin-defined under the `interface.*` namespace.
```

[cols="2,1,1,2,3,2"]
|===
| Field | Type | ECS status | SNMP MIB OID | Description | Example

| `interface.name`
| keyword
| Custom
| `ifDescr` (1.3.6.1.2.1.2.2.1.2)
| Interface name.
| `Gi0/0/0`, `eth0`, `xe-0/0/0`

| `interface.speed`
| long
| Custom
| `ifSpeed` (1.3.6.1.2.1.2.2.1.5)
| Interface speed in bits/sec.
| `10000000000` (10 Gbps)

| `interface.status.admin`
| keyword
| Custom
| `ifAdminStatus` (1.3.6.1.2.1.2.2.1.7)
| Administrative status.
| `up`, `down`

| `interface.status.oper`
| keyword
| Custom
| `ifOperStatus` (1.3.6.1.2.1.2.2.1.8)
| Operational status.
| `up`, `down`, `testing`

| `interface.traffic.in.bytes`
| long
| Custom
| `ifInOctets` (1.3.6.1.2.1.2.2.1.10)
| Cumulative inbound bytes.
| `125000000`

| `interface.traffic.out.bytes`
| long
| Custom
| `ifOutOctets` (1.3.6.1.2.1.2.2.1.16)
| Cumulative outbound bytes.
| `87500000`

| `interface.errors.in`
| long
| Custom
| `ifInErrors` (1.3.6.1.2.1.2.2.1.14)
| Inbound error count.
| `0`

| `interface.errors.out`
| long
| Custom
| `ifOutErrors` (1.3.6.1.2.1.2.2.1.20)
| Outbound error count.
| `0`
|===

> NOTE: These are cumulative counters, not rates. To compute bits/sec, take the delta between two consecutive polls in your collector or with a {kib} scripted field.

Reviewer: ```suggestion
NOTE: These are cumulative counters, not rates. To compute bits/sec, take the delta between two consecutive polls in your collector or use a {kib} scripted field.
```
Not 100% what is intended here but maybe this works

| `arp.mac_addr` present
| Layer 3 neighbor discovery.

| MAC table entry
| `mac_table.mac_addr` present
| Layer 2 forwarding topology.

| IP address entry
| `ip_addr.address` present
| Interface IPs and subnet membership for segment views.

| BGP peer session
| `bgp_peer.remote_ip` present
| BGP peering state, AS numbers, prefix counts, uptime.

| OSPF neighbor
| `ospf_neighbor.neighbor_ip` present
| OSPF adjacency state, router ID, area, priority.
|===

All six share the same `host.*`, `observer.*`, and `network.*` fields to identify which device the data belongs to.

[discrete]
[[network-topology-fields-ingest-pipeline]]
== Ingest pipeline: `snmp-device-enrichment`

The ingest pipeline performs the following on every incoming document:

. *Device type default* — sets `host.type = "unknown"` if the field is absent.
. *Site default* — sets `network.site = "Ungrouped"` if the field is absent.
. *Device type inference (fallback)* — if `host.type` is still `"unknown"`, performs a keyword match on `observer.sys_descr` to set `host.type` (router, switch, firewall, ap, server). This is a best-effort fallback; setting `host.type` explicitly in the collector config is preferred.

> . *Vendor detection* — runs a regex match on `observer.sys_descr` to set `observer.vendor`. Cisco, Juniper, Arista, Fortinet, Palo Alto, HPE, and Aruba are recognized out of the box, and you can extend the pipeline to recognize additional vendors.

Reviewer: Perhaps worth creating a follow-up issue to give more guidance about getting additional vendors recognized.
